### PR TITLE
banshee: Add support for SSR instructions

### PR DESCRIPTION
**Pull request by noah95**
This adds support for the ssr instructions
- Immediate versions `scfgri` and `scfgwi`
- Register versions `scfgr` and `scfgw`

Some changes to the parsing of DMA instructions had to be made, but all tests complete successfully. 

The `riscv.rs` file is now definitely out of sync with the generated one by `make src/riscv.rs`, that will be fixed in a later PR.

### CODE REVIEW

**Comment by noah95** [comment]
#62 fixes the out-of-sync issue

File: sw/banshee/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@
 SHELL := /bin/bash
 RISCV_OPCODES ?= ../../sw/vendor/riscv-opcodes
 PARSE_OPCODES ?= $(RISCV_OPCODES)/parse-opcodes
-OPCODES := opcodes opcodes-rep opcodes-dma
+OPCODES := opcodes opcodes-rep opcodes-dma opcodes-ssr
```
Comment (fabianschuiki) [inline sw/banshee/Makefile:8]: Nice, that got lost in the last few weeks.

File: sw/banshee/src/tran.rs [hunk 2]
```diff
@@ -1194,6 +1199,30 @@ impl<'a> InstructionTranslator<'a> {
...
+        match data.op {
+            riscv::OpcodeImm12Rs1::Scfgwi => {
+                // ssr write immediate holds address offset in imm12, content in rs1
+                let addr = LLVMBuildAdd(
+                    self.builder,
+                    LLVMConstInt(LLVMInt32Type(), SSR_BASE, 0),
+                    imm,
+                    NONAME,
+                );
+                self.write_mem(addr, rs1, 2);
```
Comment (fabianschuiki) [inline sw/banshee/src/tran.rs:1220]: That's a clever way to solve this for now! Later we'll probably want to directly call into some SSR config read/write functions here, and forward accesses inside `write_mem` to those functions as well (might already be halfway implement). But no need to do this now 👍

File: sw/banshee/src/tran.rs [hunk 3]
```diff
@@ -2537,8 +2602,8 @@ impl<'a> InstructionTranslator<'a> {
         &self,
         addr: LLVMValueRef,
     ) -> (LLVMValueRef, LLVMValueRef, LLVMValueRef) {
-        let ssr_start = LLVMConstInt(LLVMInt32Type(), 0x204800, 0);
-        let ssr_end = LLVMConstInt(LLVMInt32Type(), 0x204800 + 32 * 8 * 2, 0);
+        let ssr_start = LLVMConstInt(LLVMInt32Type(), SSR_BASE, 0);
+        let ssr_end = LLVMConstInt(LLVMInt32Type(), SSR_BASE + 32 * 8 * 2, 0);
```
Comment (fabianschuiki) [inline sw/banshee/src/tran.rs:2606]: 👍

File: sw/banshee/tests/ssr_inst_check.c [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        "scfgw  t1, t0\n"              // val, adr
+        "scfgr  t3, t0\n"              // val, adr
+        // compare
+        "beq    t1, t3, 4f\n"
+        "addi   %[errs], %[errs], 1\n"  // increment error on mismatch
+        "4:\n"
+        : [ errs ] "+r"(errs)
+        :);
+    return errs;
+}
```
Comment (fabianschuiki) [inline sw/banshee/tests/ssr_inst_check.c:56]: Very very nice, thanks for also adding a test 🎉

**Review by fabianschuiki — APPROVED** [review]
Looks good to me, thanks a lot for tackling this!

Just a minor nitpick: Maybe we don't really want to commit the `sw/vendor/riscv-opcodes/inst.rs` as it is redundant with `sw/banshee/src/riscv.rs`?

File: sw/banshee/src/tran.rs (hunk 2 above)
Comment (noah95) [reply sw/banshee/src/tran.rs:1220]: Oh you mean using `banshee_ssr_write_cfg`?

**Comment by noah95** [comment]
> Just a minor nitpick: Maybe we don't really want to commit the `sw/vendor/riscv-opcodes/inst.rs` as it is redundant with `sw/banshee/src/riscv.rs`?

Oops, didn't see that there. Removed in f7e8d57

**Review by zarubaf — APPROVED** [review]
Awesome, LGTM! @fabianschuiki ready to merge?

**Comment by fabianschuiki** [comment]
LGTM! Merging :+1:

**Pull request merged**